### PR TITLE
add ability to export/collect scrape_jobs

### LINKED ISSUE #126: Add support for exporting/collecting *_exporter configs
It would be nice to have the individual `*_exporter` export puppet resources that could be collected on the prometheus node and automatically configure scrapers.

I already have a PR for this on the pipeline, so this ticket is just for reference, while the PR doesn't get submitted.

### PR DESCRIPTION
**Pull request by anarcat**
#### Pull Request (PR) description

This is a revised version of #141, adapted to apply cleanly on master, but also simplified in that the datastructures for the scrape collector are cleaner. It's a array of strings (job names) instead of an array of hashes containing the job names and optional scheme.

I haven't tested this yet, but figured I would share my work for peer review anyways. I believe I have taken into account the comments in #141.

#### This Pull Request (PR) fixes the following issues

Fixes #126

### CODE REVIEW

File: manifests/scrape_job.pp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  Array[String] $targets,
+  Hash[String, String] $labels = {},
+  String $collect_dir          = undef,
+  Optional[String] $scheme     = 'http',
+) {
+  $config = to_yaml([
+    {
+      targets => $targets,
+      labels  => $labels,
+      scheme  => $scheme,
```
Comment (costela) [inline manifests/scrape_job.pp]: Does this actually work? Can't find the `scheme` key in [the docs](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#file_sd_config).
The reason for my original - admittedly a bit unintuitive - list of hashes approach was precisely because the scheme seems to only exist in the  `scrape_config` context, not in the context of the individual `targets` (and therefore not in the context of the puppet `scrape_job` exported resource).
Did I miss something?

**Review by costela — COMMENTED** [review]
As commented on #141: thanks for picking up the work!
I'm a bit unsure as to how the "flat-list" will work, though. See comment on code.

File: manifests/scrape_job.pp (hunk 1 above)
Comment (anarcat) [reply manifests/scrape_job.pp]: ouch. you're probably right - i'll take a look.

Comment (anarcat) [reply manifests/scrape_job.pp]: you're right, it does not work. i'll see if I can just reuse the original pattern... :/

File: manifests/daemon.pp [hunk 2]
```diff
@@ -81,6 +81,9 @@
   Boolean $manage_service         = true,
   Hash[String, Scalar] $env_vars  = {},
   Optional[String] $env_file_path = $prometheus::env_file_path,
+  Boolean $export_scrape_job      = false,
+  String $scrape_host             = $facts['fqdn'],
```
Comment (bastelfreak) [inline manifests/daemon.pp]: We can probably use this datatype? https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/fqdn.pp

File: manifests/scrape_job.pp (hunk 1 above)
Comment (anarcat) [reply manifests/scrape_job.pp]: i've reverted to the origina data structure in #141 even though i don't quite like it. :)

**Comment by anarcat** [comment]
Alright, i've redone this patch to use the original data structure from #141 even if I don't quite like it. My main problem with it is that it needs to have an explicit list of job names in a configuration somewhere. by default, it will *not* collect anything unless the operator *guesses* those job names, and those are not exactly obvious.

For example, here the node exporter job is called `prometheus-node-exporter` (note the dashes) while by default it would be `prometheus-node_exporter` (note the dash and underscore). That's because I've done some work to sync with the Debian package, which uses a dash (see #303). It won't be obvious for new users how to configure those new jobs and I would prefer if *all* jobs are collected by default but understand this is not very practical as such right now.

Besides, this works for me in production now. I suspect tests might fail (again) and there is no test coverage for the new code (and there wasn't any in #141 either). I must admit I have never worked with tests in Puppet manifests (let alone for exported resources) so I would really appreciate help and guidance to complete that if it's a requirement to getting that merged.

That said, the thing works for me as such, and I believe it would already be a huge improvement over the "no internal discovery" current situation. :)

**Comment by anarcat** [comment]
the CI build failed, but i believe it's a transient error, actually. it would be great if some admin could reschedule that...

**Comment by costela** [comment]
LGTM
the tests are unfortunately also where I got stuck for lack of time :/

File: manifests/server.pp [hunk 3]
```diff
@@ -43,6 +43,7 @@
   Boolean $purge_config_dir                                                     = $prometheus::purge_config_dir,
   Boolean $manage_user                                                          = $prometheus::manage_user,
   Optional[Variant[Stdlib::HTTPurl, Stdlib::Unixpath, String[1]]] $external_url = $prometheus::external_url,
+  Optional[Array[Hash[String, String]]] $collect_scrape_jobs                    = [],
```
Comment (bastelfreak) [inline manifests/server.pp]: since the default is an empty array, you don't need to make it optional. Also please enforce that the string is at least 1 char long:
```suggestion
  Array[Hash[String[1], String[1]]] $collect_scrape_jobs                        = [],
```

File: templates/prometheus.yaml.erb [hunk 4]
```diff
@@ -7,7 +7,7 @@
 <% full_config = {
   'global'=>global_config,
   'rule_files'=>rule_files,
-  'scrape_configs'=>scrape_configs,
+  'scrape_configs'=>scrape_configs + @collected_scrape_jobs,
```
Comment (bastelfreak) [inline templates/prometheus.yaml.erb:10]: please write a test to verify this line works,

File: manifests/apache_exporter.pp [hunk 5]
```diff
@@ -100,6 +100,7 @@
   String[1] $config_mode                                             = $prometheus::config_mode,
   String[1] $arch                                                    = $prometheus::real_arch,
   Stdlib::Absolutepath $bin_dir                                      = $prometheus::bin_dir,
+  String $scrape_job_name                                            = 'apache',
```
Comment (bastelfreak) [inline manifests/apache_exporter.pp]: please enforce the minimal string length here as well
```suggestion
  String[1] $scrape_job_name                                         = 'apache',
```

Comment (bastelfreak) [reply manifests/apache_exporter.pp]: you probably need to add the `scrape_job_name`, `scrape_port` and `scrape_port` to all exporters?

File: templates/prometheus.yaml.erb (hunk 4 above)
Comment (anarcat) [reply templates/prometheus.yaml.erb:10]: do you have any suggestions on where i should start with this? i'm not very familiar with the puppet modules test environment, and even less so when it concerns exported resources...

File: manifests/apache_exporter.pp (hunk 5 above)
Comment (anarcat) [reply manifests/apache_exporter.pp]: > you probably need to add the scrape_job_name, scrape_port and scrape_port to all exporters?

i've been just scratching an itch so far. I'm not very familiar with the other exporters so I don't know what port or name name they usually use... so i guess this could be an exercise for the reader? the point of the PR is to kickstart the collection, if there are bits missing i don't think it's that big of a deal for a first phase...

e.g. in time I might also work with the bind, postgres, and mysql exporters and can submit separate PRs for those when the time comes. i don't feel comfortable setting that for exporters I cannot test at all.

File: templates/prometheus.yaml.erb (hunk 4 above)
Comment (costela) [reply templates/prometheus.yaml.erb:10]: I'm not sure how much you know, so maybe this sounds a bit condescending, but basically you can just:
```
$ bundle install --with test
$ bundle exec rake test
```
And then just go through the stuff in the `spec` folder to get a feel for how things are currently being tested.
As for exported resources, here are some basic examples:
https://github.com/rodjek/rspec-puppet#testing-exported-resources

And yes: I know this is all easier said than done :wink:

Comment (anarcat) [reply templates/prometheus.yaml.erb:10]: it's not condescending, I had truly no idea where to start so your feedback was definitely useful. that said:

    bundle install --with test

this is the kind of horrors that make me hesitant of actually getting down that rabbit hole. i modestly imagined this would do a third-party setup of gem things in `/usr/local` but i have actually *zero* idea where all that junk it downloaded was installed or why. it *looks* like it shoved a bunch of things in `/var/lib` but I'm going to pretend all this is perfectly normal and I won't need to worry about arbitrary code being installed as root on my machine. but that's on par for the game these days...

ANYWAY. :)

i tried to do the thing, i really did. here's the patch I came up with:

```diff
modified   spec/defines/daemon_spec.rb
@@ -24,7 +24,10 @@
           group:             'smurf_group',
           env_vars:          { SOMEVAR: 42 },
           bin_dir:           '/usr/local/bin',
-          install_method:    'url'
+          install_method:    'url',
+          export_scrape_job: true,
+          scrape_host:       'localhost',
+          scrape_port:       1234,
         }
       ].each do |parameters|
         context "with parameters #{parameters}" do
@@ -229,6 +232,11 @@
               'enable' => true
             )
           }
+          context 'exported resources' do
+            subject { exported_resources }
+
+            it { is_expected.to contain_file('/etc/prometheus/file_sd_config.d/smurf_exporter_localhost:1234.yaml') }
+          end
         end
       end
     end

```

this fails with:

```
     Failure/Error: it { is_expected.to contain_file('/etc/prometheus/file_sd_config.d/smurf_exporter_localhost:1234.yaml') }
       expected that the catalogue would contain File[/etc/prometheus/file_sd_config.d/smurf_exporter_localhost:1234.yaml]
     # ./spec/defines/daemon_spec.rb:238:in `block (7 levels) in <top (required)>'
```

so the file isn't created? did i do a typo or guessed the name wrong? how do i tell?

... i would love to write test, I really would. but this is just incomprehensible to me. :)

File: manifests/config.pp [hunk 6]
```diff
@@ -110,6 +110,39 @@
...
+    }
+
+    $job_name = $job_definition['job_name']
+
+    Prometheus::Scrape_job <<| job_name == $job_name |>> {
+      collect_dir => "${prometheus::config_dir}/file_sd_config.d",
+      notify      => Class['::prometheus::service_reload'],
+    }
+  }
+  # collect the scrape jobs in a single list that gets appended to
```
Comment (costela) [inline manifests/config.pp]: nitpicking: "collect" here is a bit ambiguous, because it is also used to refer to puppet's export/collect mechanic. Maybe rephrase as "join the scape jobs (...)"?

File: templates/prometheus.yaml.erb (hunk 4 above)
Comment (costela) [reply templates/prometheus.yaml.erb:10]: The bundle stuff should all go inside the `vendor` folder inside the project. AFAIK it shouldn't install anything system-wide. You mentioned "installed as root" and maybe that's part of the misunderstanding: `bundle` should be called as your normal non-root user. That's what the `$` at the beginning of the commands is meant to indicate. It's a very old convention: `$` indicates the normal user prompt, `#` indicates a root prompt. Just in case you didn't know (again: apologies if this comes over as condescending)

As for the actual test: the `contain_*` part is used to test the existence of some resource in the generated manifest. In this case the exported resource. The file you are testing for is not exported, it is realized by the exported resource, namely `scrape_job`. So here you should be testing if `scrape_job` is contained.
Does this help?

Comment (bastelfreak) [reply templates/prometheus.yaml.erb:10]: I recommend our CONTRIBUTING.md: https://github.com/voxpupuli/puppet-prometheus/blob/master/.github/CONTRIBUTING.md

`bundle install` doesn't install anything as root, unless you prefix it with sudo. But our recommendation is the option `--path .vendor/` which will install all the gems into a subdir within the current module and won't pollute $PATH.

Comment (anarcat) [reply templates/prometheus.yaml.erb:10]: i know about `#` and `$`. :) i did run `bundle install` as regular user, but it said it couldn't write to `/var/lib` and asked me for `sudo`... anyways, it turns out it all installed stuff in `/var/lib/gems-2.5.0` which i believe was just empty before (and if it wasn't, it should be). so I did the `--path` trick (thanks for that one, it looks like I hadn't done my homework!)... and I got this error:

```
Downloading listen-3.1.5 revealed dependencies not in the API or the lockfile
(rb-fsevent (>= 0.9.4, ~> 0.9), ruby_dep (~> 1.2)).
Either installing with `--full-index` or running `bundle update listen` should
fix the problem.
```

which i didn't get without `--path`. running with `--full-index` doesn't fix the problem, but `bundle update listen` does, so I plowed ahead.

my tests are still failing. I tried this instead:

```
            it { is_expected.to contain_scrape_job('localhost:1234') }
```

and also:

```
            it { is_expected.to contain_prometheus_scrape_job('localhost:1234') }
```

they both fail with an error similar to the original:

```
     Failure/Error: it { is_expected.to contain_prometheus_scrape_job('localhost:1234') }
       expected that the catalogue would contain Prometheus_scrape_job[localhost:1234]
     # ./spec/defines/daemon_spec.rb:238:in `block (7 levels) in <top (required)>'
```

<del>is there a way to list the resources that are actually exported?</del> where are those tests running anyways? how can I debug this?

thanks for the hand holding! :)

update: it does look like the resource gets exported.

oh, and it *does* look like that scrape_job is exported, for what it's worth. at the end of the test run, I see:

```
[...]
Untouched resources:
[...]

  Prometheus::Scrape_job[localhost:1234]
```

so it looks like I'm messing up the quoting or something on that front...

... and worse, it doesn't address the original review here, which is:

```
    'scrape_configs'=>scrape_configs + @collected_scrape_jobs,
```

> please write a test to verify this line works,

:)

Comment (costela) [reply templates/prometheus.yaml.erb:10]: For `contain_*` examples, see [here](https://rspec-puppet.com/documentation/defined_types/#test-a-resource). In this specific case you should probably be using `contain_prometheus__scrape_job` (note the "dunder"/double underscore to denote "::").

As for @bastelfreak 's original comment: yes, you'd have to also test the contents of the generated file to be sure the concatenation makes sense. This should be done in a separate test case, though.

File: manifests/config.pp (hunk 6 above)
Comment (anarcat) [reply manifests/config.pp]: i used "assemble"

File: templates/prometheus.yaml.erb (hunk 4 above)
Comment (anarcat) [reply templates/prometheus.yaml.erb:10]: alright, i pushed what i think is a working test case for `scrape_job.pp`, let me know if that looks alright to you:

```ruby
          context 'exported resources' do
            subject { exported_resources }
            it {
              is_expected.to contain_prometheus__scrape_job('localhost:1234')
            }
          end
```

also pushed here...

I did try to test the original line commented here, but I couldn't figure out how to create the scrape_job resources (the daemon, really) from the other file. I tried to add this to `classes/prometheus_rspec.rb`:

```ruby
      context 'exported resources' do
        [
          {
            collect_scrape_jobs: [
              { job_name: 'smurf_exporter' },
            ]
          }
        ].each do |parameters|
          context "with prometheus version #{parameters[:version]}" do
            let(:params) do
              parameters
            end
            # copied over from defines/daemon_spec.rb
            let(:pre_condition) { 'prometheus::daemon { "smurf_exporter": export_scrape_job => true, scrape_host => "localhost", scrape_port => 1234, real_download_url => "https://github.com/prometheus/smurf_exporter/releases/v1.2.3/smurf_exporter-1.2.3.any.tar.gz"' }
            subject { exported_resources }
            it {
              is_expected.to contain_prometheus__scrape_job('localhost:1234')
            }
          end
        end
      end
```

but that crashes with:

```
  6) prometheus on ubuntu-18.04-x86_64 exported resources with prometheus version  should contain Prometheus::Scrape_job[localhost:1234]
     Failure/Error: is_expected.to contain_prometheus__scrape_job('localhost:1234')
     
     Puppet::ParseErrorWithIssue:
       Could not parse for environment rp_env: Syntax error at 'class' (line: 3, column: 1) on node curie.anarc.at
```

... since i'm just cargo culting this, i have no idea why it fails again. ;) 

i guess that rspec is creating a puppet manifest and applying it with that stuff - is there a place where I can actually see what manifest it's creating to debug what i'm doing wrong?

thanks again for all the hand holding! we're getting closer!

**Review by costela — CHANGES_REQUESTED** [review]
Sorry for the delayed answer.
I've started using your branch to check it out and try to work on the specs, and I've found a few other things to clear up.

If you agree with my comments, you might want to cherry pick the two top commits from our RegioHelden/puppet-prometheus

File: manifests/node_exporter.pp [hunk 7]
```diff
@@ -112,6 +112,7 @@
   Optional[Array[String]] $collectors = undef,
   Array[String] $collectors_enable    = [],
   Array[String] $collectors_disable   = [],
+  String $scrape_job_name             = 'node',
```
[the remark this answers is not in the document]
Comment (anarcat) [reply manifests/node_exporter.pp]: the problem is that traditionnally, `node` is used in job names everywhere in Prometheus, not `node_exporter`. and we don't have the actual `node` string anywhere in the module that we could use, so we do need to pass it along here somehow. we *could* hardcode it to `node` within the exporter class, but i figured it was a nice touch to allow overriding it because it *does* get changed sometimes... :/

File: manifests/init.pp [hunk 8]
```diff
@@ -180,6 +188,7 @@
   Hash $config_defaults = {},
   String $os            = downcase($facts['kernel']),
   Optional[Variant[Stdlib::HTTPUrl, Stdlib::Unixpath, String[1]]] $external_url = undef,
+  Optional[Array[Hash[String, String]]] $collect_scrape_jobs = [],
```
[the remark this answers is not in the document]
Comment (anarcat) [reply manifests/init.pp]: sure, sounds reasonable.

File: manifests/server.pp [hunk 9]
```diff
@@ -43,6 +43,7 @@
   Boolean $purge_config_dir                                                     = $prometheus::purge_config_dir,
   Boolean $manage_user                                                          = $prometheus::manage_user,
   Optional[Variant[Stdlib::HTTPurl, Stdlib::Unixpath, String[1]]] $external_url = $prometheus::external_url,
+  Array[Hash[String[1], String[1]]] $collect_scrape_jobs                        = [],
```
[the remark this answers is not in the document]
Comment (anarcat) [reply manifests/server.pp]: ah yes, quite possibly.

File: manifests/init.pp (hunk 8 above)
[the remark this answers is not in the document]
Comment (anarcat) [reply manifests/init.pp]: better?

File: manifests/server.pp (hunk 9 above)
[the remark this answers is not in the document]
Comment (anarcat) [reply manifests/server.pp]: better?

**Comment by anarcat** [comment]
> If you agree with my comments, you might want to cherry pick the two top commits from our RegioHelden/puppet-prometheus

I've reviewed your comments individually, but couldn't figure out which two commits you were refering to here. Maybe I just couldn't find them but I suspect you forgot to push. Mind pasting the actual commit hashes to give me a hand here?

File: manifests/node_exporter.pp (hunk 7 above)
[the remark this answers is not in the document]
Comment (costela) [reply manifests/node_exporter.pp]: I'm not sure this prometheus naming tradition should play a role here. If we're managing prometheus using puppet, it's IMHO more useful to keep the cognitive load small by adhering to a single naming convention inside of the context where they are used (puppet), than adhering to a tradition from an outside context (prometheus).
After all, I suspect the average puppet-prometheus user will not be interested in the prometheus naming conventions. They will use `include ::prometheus::node_exporter` and will wonder why they then have to use `prometheus::scrape_jobs: - job_name: node (...)`.
But this just a gut feeling, of course.

**Comment by costela** [comment]
@anarcat I mean the top two commits here: https://github.com/RegioHelden/puppet-prometheus/commits/1c9220b85708670aa80eba5943b3fd15fde75a97
The main addition in these commits are exporter settings for all available `*_exporter.pp`.

File: manifests/node_exporter.pp (hunk 7 above)
[the remark this answers is not in the document]
Comment (anarcat) [reply manifests/node_exporter.pp]: > After all, I suspect the average puppet-prometheus user will not be interested in the prometheus naming conventions.

Oh hell yeah they will. :) At least I am: the next step after setting up those exporters for me was to set up Grafana dashboards and *those* expected the job name to be `node`, not `node_exporter`. Rewriting all those dashboards would have been a massive pain in the bottom and it was much *MUCH* easier for me to patch the module this way. :)

**Comment by anarcat** [comment]
okay, in there i think i got b71f2225aa7adfaa292daa6283f538f0b62bf030 covered. 1c9220b85708670aa80eba5943b3fd15fde75a97 includes a bunch of stuff including the disagreement we have about removing `scrape_job_name`, so obviously I tend to disregard that part. but the chunks about the port numbers do seem quite useful.

and about that, i have been thinking about DRY here: i had to copy-paste those port numbers in my firewall rules elsewhere and thought of how that could be repeated, but couldn't find a clean solution... just something to keep in mind i guess...

File: manifests/node_exporter.pp (hunk 7 above)
[the remark this answers is not in the document]
Comment (costela) [reply manifests/node_exporter.pp]: :thinking: hm, you might have a point.
Mismatch still feels wrong, IMHO, but since I can't suggest a better way to deal with it, I have to concede this one.

**Comment by anarcat** [comment]
so i think i resolved all the review comments here and finally fixed the test suite so it's all green.

i understand there might be future improvements to be done on top of this PR, but that's just it: it can be done on top! i think we should go ahead and merge this massive commit before it goes stale. then extra goodies can be slapped on top in other PRs.

how does that sound?

**Comment by anarcat** [comment]
i have rebased this, and i think it's ready to be merged. can someone approve and merge this?

File: manifests/scrape_job.pp [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+#
+#  [*collect_dir*]
+#  Directory used for collecting scrape definitions.
+#  NOTE: this is a prometheus setting and will be overridden during collection.
+#
+define prometheus::scrape_job (
+  String $job_name,
+  Array[String] $targets,
+  Hash[String, String] $labels = {},
+  String $collect_dir          = undef,
```
Comment (bastelfreak) [inline manifests/scrape_job.pp]: can you please use `Stdlib::Absolutepath`?

Comment (bastelfreak) [inline manifests/scrape_job.pp]: please enforce the minimal string length for this type and the one above with `String[1]`

**Comment by costela** [comment]
@bastelfreak PTAL

**Comment by bastelfreak** [comment]
Okay, I got a few questions:
* Why did you add the `$scrape_job_name` parameter to some exporters but not all?
* Most exporters don't have the `$scrape_port` parameter, on some you hardcoded a port so it isn't possible to change it. Is there a reason for this? This seems to be very inconsistent
* Same applies for the `$export_scrape_job` parameter
* Also I think `$export_scrape_job` should default to false everywhere, to keep this backwards compatible. Otherwise it would be a breaking change

**Comment by anarcat** [comment]
> Why did you add the $scrape_job_name parameter to some exporters but not all?

Selfishness. :) Those are the exporters I used and wanted to avoid divergence with the manual config I had already done. Plus those match what the published Grafana dashboards expect as job names. I don't know about the exporters so I didn't touch those.

We discussed this before, by the way: https://github.com/voxpupuli/puppet-prometheus/pull/304#discussion_r267135148 and https://github.com/voxpupuli/puppet-prometheus/pull/304#discussion_r273046215

My position then was that I wanted to scratch my own itch and adding "export support" to the other exporters could be done in another, later PR.

> Most exporters don't have the $scrape_port parameter, on some you hardcoded a port so it isn't possible to change it. Is there a reason for this? This seems to be very inconsistent

Again, scratching an itch... The exporters do have default port, and they match the ports specified here. Prometheus has a semi-static list of such ports documented in their wiki and it is quite unlikely people would gratuitously change this. I wanted to avoid yet another parameter that gets trickled up the entire stack again.

> Same applies for the $export_scrape_job parameter

I'm going to assume this applies to the first comment on `$scrape_job_name` and a similar response applies.

> Also I think $export_scrape_job should default to false everywhere, to keep this backwards compatible. Otherwise it would be a breaking change

that's a good point, I can do that.

**Comment by costela** [comment]
>     * Most exporters don't have the `$scrape_port` parameter, on some you hardcoded a port so it isn't possible to change it. Is there a reason for this? This seems to be very inconsistent

@anarcat I had a commit adding the parameter to all supported exporters. It came together with other changes you argued convincingly against, but do you remember why you didn't take this specific part?
We could also easily turn them into parameters. It's not like it's a big change :wink:

**Comment by anarcat** [comment]
@costela honestly, i don't remember. it's been months now...

**Comment by costela** [comment]
@anarcat please check my PR to your fork at anarcat/puppet-prometheus#1
AFAICS it should address all of @bastelfreak 's latest comments.

**Comment by anarcat** [comment]
awesome thanks, merged!

**Review by bastelfreak — APPROVED** [review]
looks good to me. I will leave this open a few days if anybody else wants to review this

File: manifests/apache_exporter.pp [hunk 11]
```diff
@@ -133,5 +135,8 @@
     service_ensure     => $service_ensure,
     service_enable     => $service_enable,
     manage_service     => $manage_service,
+    export_scrape_job  => false,
```
Comment (alexjfisher) [inline manifests/apache_exporter.pp]: Should this be `export_scrape_job => $export_scrape_job` with `Boolean $export_scrape_job = false,` added to the class parameters?

Comment (costela) [reply manifests/apache_exporter.pp]: oh, jeez, yes! that's embarassing... :see_no_evil:

**Review by alexjfisher — CHANGES_REQUESTED** [review]
Need to fix inability to set export_scrape_job to `true`

**Comment by costela** [comment]
@alexjfisher I'll get to it in about one hour, unless @anarcat beats me to it ;)

**Comment by costela** [comment]
@anarcat can you please accept anarcat/puppet-prometheus#2? Should fix the remaining open point.

**Comment by anarcat** [comment]
done. what's the policy on PRs again? should i squash all of this in one commit? or rebase in a cleaner set? it has gotten a bit messy...

**Comment by costela** [comment]
@anarcat it would be nice if we could keep at least 2 commits, one for each of us :wink:
But if that's against voxpupuli's policy, than feel free to squash it down to one, after all, you carried the torch to the finish line! :+1:

**Comment by anarcat** [comment]
@costela are you okay with how the first commit shows up? you could be author and i'm committer or something... see 0712150d0cc3efb5f6beed3094bcd47e07ae993f

**Comment by costela** [comment]
@anarcat I think that wouldn't work, because whoever approves the PR and merges it will appear as "committer" (see the current master commit log for a few such cases).

But hey, this is some world-class bike-shedding we're doing! :sweat_smile:
Let's just get it merged. I'm ok with not getting attribution, after all I basically left my original PR to rot...

**Comment by anarcat** [comment]
i see okay, i'll craft beautiful snowflakes then :)

**Comment by anarcat** [comment]
alright, i merged it down to 6 commits, two (and the largest ones!) credited to you, @costela. i hope this is good for everyone!

@alexjfisher i think the latest changes address your latest review.

**Review by alexjfisher — APPROVED** [review]

**Comment by alexjfisher** [comment]
@anarcat @costela Thanks for your efforts and great team work!

**Comment by benedikt-haug** [comment]
I would absolutely love to use that feature. Do you already know when a new release will happen?

Thank you!
